### ISSUE TITLE
Opened layers control list gets covered by other controls

### ISSUE THREAD

**Issue by ppete2** [Leaflet/Leaflet#7944]
I positioned the layers control L.control.layers to _position: 'topright'_
After I open the control by hovering the mouse over it, the list of available layers overlapps other controls which are positioned by _position: 'bottomright'_
The problem: These "bottomright"-controls cover the bottom of the layers control list which makes it difficult to read or select layer entries on bottom of the list. Probably the reason is that "bottomright"-controls have a higher z-index than "topright"-controls i.e. covering them in case of overlapping.

Could you please fix this problem, cause an opened layers control should always completely cover each other control beneath. 
Or do you know another method to define the z-index order of control buttons in case they overlap?

**Comment by Malvoz** [comment]
This sounds like an app-specific issue. There's no way to tell what content developers will use in the different containers so setting higher z-index on one container may break things for other use cases? In your case you can probably just change the CSS to `.leaflet-top.leaflet-right { z-index: 1001; }`?

**Comment by IvanSanchez** [comment]
Can you publish a fiddle/codepen/[plunkr](https://leafletjs.com/edit) that displays this behaviour?

**Comment by Falke-Design** [comment]
![grafik](https://user-images.githubusercontent.com/19800037/151024765-ef3b8e6e-ee17-429d-9c1c-23d2aed3a3d8.png)

https://plnkr.co/edit/YWFY3lhbFG7OOEU7

**Comment by IvanSanchez** [comment]
I see.

The problem here is that each of the four control containers (one per corner) is its own stacking context, so changing the z-index on an individual control and hoping that it will have an effect over individual controls of other corner will only lead to disillusionment.

That also means that it's not possible to change the z-index depending on whether the layers control is collapsed or not (well, it's possible, but won't do nothing).

The only approach available is to change the z-index of the four corner control containers.... but then, the opposite case becomes a problem: things from the upper corners overlap things from the lower corners.

I don't think a solution exists for the general case. The best I can do right now is say: work around this with `.leaflet-top.leaflet-right { z-index:1001; }`.

**Comment by Falke-Design** [comment]
What if we set per default the top control layers to a higher z-index as the bottom ones? But would be a breaking change

**Comment by Malvoz** [comment]
@Falke-Design as Ivan stated:

> but then, the opposite case becomes a problem: things from the upper corners overlap things from the lower corners.

So changing the default z-index _may_ provide a better default, but will not solve the issue for all use cases.

**Comment by Falke-Design** [comment]
Yes this is clear but maybe it would make little bit more sense, because of: _top is higher so it is above the bottom container_ or _the most user interaction controls are in the top container_

A other solution would be to create a "interaction"-css-class which will be applied to the container when the control is clicked or `mouseover` event is fired. Additionally we could add a flag, if set the control is a user interaction control and new CSS-class will be applied and the control is not like the scale / attribution control without interaction and no need to change the z-index.

**Comment by ppete2** [comment]
The simple solution of @IvanSanchez by changing z-index works! :-)

IMHO it would make sense to allocate the top control layers a higher z-index as the bottom ones. Of course this would not solve each of possible cover-issues. But the main issue (popular used layers control list - which probably in most of the user cases is placed soewhere in the top - gets covered by bottom controls) would be solved.

Other ideas like the mentioned "interaction"-class or new options for defining z-indexes for individual controls could be implemented later too. But first the interchange of z-indexes of top and bottom corner control containers in Leaflet's default would solve the problem with Leaflet's layer control.

**Comment by IvanSanchez** [comment]
Actually.... `.leaflet-control-container > div:hover { z-index: 1001; }` might just work. Instead of the "interaction" CSS class, the `:hover` pseudo-selector should do the trick.

Can you folks see if this approach makes sense (and works cross-browser)?

It might be tricky for the touchscreen use cases, tho. :thinking:

**Comment by Malvoz** [comment]
There's a related issue: https://github.com/Leaflet/Leaflet/issues/4644.

**Comment by Malvoz** [comment]
> `.leaflet-control-container > div:hover { z-index: 1001; }` might just work.

`:hover` could be accompanied by [`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within). It would at least ensure that the components are visible when interacted with. There could still be issues on touch though, I guess.

**Comment by Malvoz** [comment]
> > `.leaflet-control-container > div:hover { z-index: 1001; }` might just work.
>
> `:hover` could be accompanied by [`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within).

Test case: https://plnkr.co/edit/1Hl2n2i71oxJzLvn

> There could still be issues on touch though, I guess.

The problem is that if the layer control is completely covered by something, it can't be interacted with, and as such the user wont be able to uncover it. The only robust solution is to:

> change the CSS to `.leaflet-top.leaflet-right { z-index: 1001; }`

but that assumes the layer control is always in the top right corner...
